### PR TITLE
Introduce Basic Staging Capabilities Utilizing the Policy Staging Ref

### PR DESCRIPTION
**Pull request by neilnaveen**
**How does the policy staging currently work** 

1. All changes to the policy are first made in the policy staging ref. These changes only impact the repo once they are merged into the policy ref, utilizing a fast-forward merge.
2. Before merging anything from the policy staging ref into the policy ref, the latest state of the policy staging ref is inspected and checked to see if it is a valid state. This guarantees that the project will always work with a good policy state since the policy ref leapfrogs from a valid state to a valid state. 
3. This is especially useful when attestations are added into gittuf, meaning the state would have to be invalid at specific points before being valid. 
4. For example, let's say that we want to have three root key holders in our root of trust from day one. We also want to set the threshold to three. Without policy staging, if one key is added, but the threshold is 3, this would be impossible to do. But if this was in the policy-staging ref, these changes could be pushed up the upstream repo, fetched by other key holders, and added to the root of trust. Finally, the changes can be merged into the main when all three keys are added.

**Possible features for policy staging** 

1. Having only one policy staging reference is limiting since they can be overwritten, so instead of having a single policy staging reference, you could potentially create multiple feature policy "branch" references.

**Problems that are created by policy staging**

1. We cannot assume that the first RSL entry is the first valid state since it most likely would be a policy-staging RSL entry, which could mean it could be invalid. That logic will have to be changed.
2. Currently, if you only have one policy staging branch, this would mean that whenever you pull the latest RSL from upstream, your WIP policy changes would be overwritten.

### CODE REVIEW

**Comment by neilnaveen** [comment]
listr-ules is being modified only for the purpose of seeing the difference between the policy staging ref and the policy ref.

File: internal/policy/policy.go
Comment (adityasaky) [inline internal/policy/policy.go:753]: gitinterface.KnowsCommit does this for us IIRC

File: internal/policy/policy.go
Comment (adityasaky) [reply internal/policy/policy.go:503]: The question of workflow would then extend to the repository package which is where policy is actually updated.

**Review by adityasaky — COMMENTED** [review]
Added some thoughts. This might be an issue that warrants a synchronous discussion of workflows. cc @JustinCappos @wlynch @reza-curtmola

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go]: Fixed

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go]: Fixed

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go]: Fixed

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go]: I have fixed this. So state.Commit will never be used in a way to change the policy ref. Does this mean that the policy ref is almost useless, other than the fact that it's RSL entries are only on valid states that where applied?

**Comment by neilnaveen** [comment]
Rebased the code with main 👀 @adityasaky

File: internal/policy/policy.go
Comment (adityasaky) [inline internal/policy/policy.go:881]:
```suggestion

	policyCommit, err := gitinterface.GetCommit(repo, policyRef.Hash())
```

File: internal/policy/policy.go
Comment (adityasaky) [inline internal/policy/policy.go:894]: can we use a named error here?

File: internal/policy/policy_test.go
Comment (adityasaky) [inline internal/policy/policy_test.go:536]: I'd like to see more cases, split up with `t.Run` that also inspects the policy and policy-staging refs post apply.

File: internal/policy/verify.go
Comment (adityasaky) [inline internal/policy/verify.go:172]: No, we should never be loading the policy staging ref during verification.

Comment (adityasaky) [reply internal/policy/verify.go:172]: staging should be special cased so we skip to the next one.

File: internal/repository/policy.go
Comment (adityasaky) [inline internal/repository/policy.go:25]: We should think about the interface for Push so we know what ref is being pushed. Perhaps we don't want to push both everytime...

File: internal/repository/sync_test.go
Comment (adityasaky) [inline internal/repository/sync_test.go:88]: nit: unrelated diff

File: internal/repository/sync_test.go
Comment (adityasaky) [inline internal/repository/sync_test.go:205]: nit: unrelated diff

File: internal/repository/targets.go
Comment (adityasaky) [inline internal/repository/targets.go:12]: not sure why there's a diff here, gofmt / goimports should take care of this?

File: internal/repository/targets_test.go
Comment (adityasaky) [inline internal/repository/targets_test.go:86]: nit: newline removed

File: internal/repository/verify_test.go
Comment (adityasaky) [inline internal/repository/verify_test.go:31]: shouldn't this happen in the createTestRepositoryWithPolicy helper?

Comment (adityasaky) [reply internal/repository/verify_test.go:31]: I see it there as well.

**Review by adityasaky — CHANGES_REQUESTED** [review]
This is looking pretty good. Some thoughts in addition to my inline comments:
1) could you update the description with the new workflow for updating gittuf policy? I think we need to discuss this as a group, @JustinCappos, @wlynch.
2) we need to think about how this relates to policy state self-verification, especially with the changes in #308.

File: internal/rsl/rsl.go
Comment (neilnaveen) [reply internal/rsl/rsl.go]: Fixed

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go:613]: I thought that since all state changes should be routed through the policy staging namespace, there would be no need to commit changes to the policy namespace.

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go]: I have changed it to a required string parameter, not the whole ref name, but the end name, like `gittuf list rules -targetRef policy`

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go:862]: Added

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go:863]: I was using LoadCurrentState since it also verifies the state of the repo, which means that if this fails, apply should not be run, since verification failed on the policy staging ref

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go:888]: fixed

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go]: Fixed

File: internal/policy/policy_test.go
Comment (neilnaveen) [reply internal/policy/policy_test.go:536]: I have inspected the policy and policy-staging refs, but I do not know how to create an invalid state. How should I do so?

File: internal/policy/verify.go
Comment (neilnaveen) [reply internal/policy/verify.go]: Fixed

File: internal/repository/policy.go
Comment (neilnaveen) [reply internal/repository/policy.go:25]: Should this be a later PR, or should this be incorporated into this one as a different function in cmd/rsl

File: internal/repository/sync_test.go
Comment (neilnaveen) [reply internal/repository/sync_test.go]: Fixed

File: internal/repository/sync_test.go
Comment (neilnaveen) [reply internal/repository/sync_test.go]: Fixed

File: internal/repository/verify_test.go
Comment (neilnaveen) [reply internal/repository/verify_test.go]: Removed

File: internal/policy/policy.go
Comment (adityasaky) [reply internal/policy/policy.go:613]: I think it depends on the workflow we collectively decide on, so you're right. I'd like to hear from other maintainers here.

File: internal/policy/policy.go
Comment (adityasaky) [reply internal/policy/policy.go:863]: I see. That's a good call, let's keep this thread open, especially in the context of #308.

File: internal/policy/verify.go
Comment (adityasaky) [inline internal/policy/verify.go:172]: Move this check above the debug statement and add a debug message saying checking if it's for policy-staging?

File: internal/repository/helpers_test.go
Comment (adityasaky) [inline internal/repository/helpers_test.go:104]:
```suggestion

	if err := policy.Apply(testCtx, r.r, false); err != nil {
```

File: internal/repository/policy.go
Comment (adityasaky) [reply internal/repository/policy.go:25]: This is another what-should-the-workflow-be question, so let's see.

File: internal/policy/policy.go
Comment (adityasaky) [inline internal/policy/policy.go:880]: To fail early, we should also be performing this check when loading the staging area as we may be mutating the policy...

File: internal/rsl/rsl.go
Comment (adityasaky) [inline internal/rsl/rsl.go:31]: why is this necessary in the rsl package?

Comment (neilnaveen) [reply internal/rsl/rsl.go]: Removed

File: internal/repository/helpers_test.go
Comment (neilnaveen) [reply internal/repository/helpers_test.go:105]: Fixed

File: internal/policy/verify.go
Comment (neilnaveen) [reply internal/policy/verify.go:172]: Fixed

File: internal/cmd/policy/listrules/listrules.go
Comment (neilnaveen) [reply internal/cmd/policy/listrules/listrules.go]: I have moved this to repository/policy.

**Comment by JustinCappos** [comment]
FYI: I think we should also think about having the ability to stage and merge changes to different policy files independently.

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go:613]: I agree that the current version of policy staging is quite painful to use, especially with multiple contributors creating changes to the policy state. 

I was thinking about addressing these issues, and https://github.com/gittuf/gittuf/pull/260#issuecomment-1962047511,  in a follow-up PR since if we can merge this simple, bare-bones version of policy-staging, making it just work(albeit in a clunky manner), other problems can start to be fixed, like those listed here https://github.com/gittuf/gittuf/issues/230#issuecomment-1961448955.

**Comment by neilnaveen** [comment]
@adityasaky @JustinCappos Does each RSL entry only change one policy file at a time, since if not, staging policy files seperatly would be impossible, since we could not stage a commit that changes both file `a` and `b` if we only want commits that change file `a`

File: internal/policy/policy.go
Comment (adityasaky) [inline internal/policy/policy.go:156]: I think this is an incorrect rebase after #308? That one makes this a private function.

File: internal/cmd/policy/listrules/listrules.go
Comment (adityasaky) [inline internal/cmd/policy/listrules/listrules.go:24]: does this need to be required? Maybe make the default the main policy ref?

File: internal/policy/policy.go
Comment (adityasaky) [reply internal/policy/policy.go:663]: I think treating staging refs as any other branch is fine. The only thing I want us to think through is what happens when merging multiple policy changes. If we have a good story there, we should be fine.

File: internal/policy/verify.go
Comment (adityasaky) [inline internal/policy/verify.go:198]:
```suggestion
			}

```

File: internal/repository/policy.go
Comment (adityasaky) [inline internal/repository/policy.go:28]: We should also check what the impact of PolicyRef being up to date here is. It shouldn't assume the other refs are updated. I suspect the behavior is right but double checking / testing would be good.

File: internal/cmd/policy/listrules/listrules.go
Comment (neilnaveen) [reply internal/cmd/policy/listrules/listrules.go]: Fixed

File: internal/policy/verify.go
Comment (neilnaveen) [reply internal/policy/verify.go:198]: Fixed

File: internal/repository/policy.go
Comment (neilnaveen) [reply internal/repository/policy.go:28]: I have added a test

File: internal/policy/policy.go
Comment (neilnaveen) [reply internal/policy/policy.go]: Fixed

File: docs/cli/gittuf.md
Comment (adityasaky) [inline docs/cli/gittuf.md:20]: Looks like a mistake but this command is plumbed through as gittuf gittuf apply?

Comment (neilnaveen) [reply docs/cli/gittuf.md]: I have fixed this, but that is what `make generate`. After these changes if I run `make generate`, the changes are reversed and the `gittuf gittuf` command is added to the docs

Comment (neilnaveen) [reply docs/cli/gittuf.md]: Docs are failing, since generate creates a different set of docs with the `gittuf gittuf` command

**Comment by neilnaveen** [comment]
I have rolled back the changes to the simple version of policy staging

**Comment by adityasaky** [comment]
@patzielinski: I tried out the policy-staging for serial signing. Can you build off this branch and kick the tyres some more? I primarily stayed on the happy path FWIW.

@neilnaveen , @wlynch: I've added some helper commands such as the ability to just sign the root metadata as well as bump the root role's threshold. Neil, can you take a look at my commits? Billy, I think the whole PR could use a review by you.

Also, this functionality is still marked developer-mode only, which I think is fine to keep as is for now.

File: internal/cmd/policy/listrules/listrules.go
Comment (adityasaky) [reply internal/cmd/policy/listrules/listrules.go:24]: FWIW, I still see this in the latest commit. It possibly snuck back in? :)

**Comment by neilnaveen** [comment]
I think the commits look good, and having the functionality under developer mode also seems to be a good idea, since should not be a regularly used feature at the moment

File: internal/cmd/policy/listrules/listrules.go
Comment (neilnaveen) [reply internal/cmd/policy/listrules/listrules.go]: Fixed

File: internal/policy/policy.go
Comment (wlynch) [inline internal/policy/policy.go:75]: nit: you can inline this in the loop above:

```go
for _, name := range []string{PolicyRef, PolicyStagingRef} {
	refname := plumbing.ReferenceName(name)
	if ref, err := repo.Reference(refname, true); err != nil {
		if !errors.Is(err, plumbing.ErrReferenceNotFound) {
			return err
		}
	} else if !ref.Hash().IsZero() {
		return ErrPolicyExists
	}
	if err := repo.Storer.SetReference(plumbing.NewHashReference(refname, plumbing.ZeroHash)); err != nil {
		return err
	}
}
return nil
```

**Review by wlynch — COMMENTED** [review]
I think this needs to be rebased - there are a bunch of new commands included here that look to be unrelated to this change.

Only looking at internal, this LGTM. I'm not 100% confident in my ability to grok all the subtleties here though, so will defer to @adityasaky for final confirmation.

**Comment by adityasaky** [comment]
> there are a bunch of new commands included here that look to be unrelated to this change.

yep! I added some commands to allow users to add their extra signatures, which is one of the primary motivations to have a staging area (add root.json with all root keys, sign serially, then apply). This did muck up the PR a bit but it allowed for IRL testing of the staging bits.

**Comment by adityasaky** [comment]
@wlynch could you take a look at https://github.com/gittuf/gittuf/pull/260/commits/1c6c2a5088682cc443c35aebfdde28d846e13550 and https://github.com/gittuf/gittuf/pull/260/commits/93da1fc46e185efec141cd198bb8b41903ca8809? I noticed that root threshold wasn't being verified on apply and realized we needed to handle a couple more edge cases with apply. In my playground, I'm seeing expected behavior both in success and failure cases.

**Review by wlynch — APPROVED** [review]

**Review by adityasaky — APPROVED** [review]

**Pull request merged**
